ingestion-edge/ingestion_edge/disk_queue.py

```diff
@@ -6,65 +6,93 @@
+            ujson.dump(publish_kwargs, q)
+            q.write("\n")
             q.flush()
 
 
 @disk_queue.route("/__flush__")
 def flush_queue() -> Tuple[str, int]:
-    if not exists(QUEUE_FILE):
-        # early return for empty queue
-        return "", 204
+    published = 0
+    # read_lock allows one flush at a time to avoid duplicate publish calls
     with read_lock:
         # initialize queue state
+        try:
+            with open(QUEUE_POSITION_FILE, "rb") as qp:
+                start = int.from_bytes(qp.read(), sys.byteorder)
+        except FileNotFoundError:
+            start = 0
         done = False
-        start = 0
-        end = 0
-        if exists(QUEUE_POSITION_FILE):
-            with open(QUEUE_POSITION_FILE, "r") as qp:
-                start = int(qp.read())
+        end = start
         try:
-            # process the queue
-            # TODO release write lock between line reads
-            with write_lock:
-                # TODO rotate queue file and handle rotated files first
-                with open(QUEUE_FILE, "r") as q:
-                    for line in q:
-                        if end < start:
-                            # TODO use q.seek(start)
-                            end += 1
-                            continue
-                        client.publish(**ujson.loads(line)).result()
-                        # update end only on success
-                        end += 1
-            remove(QUEUE_FILE)
-            done = True
+            # process queue
+            # TODO rotate queue file and handle rotated files first
+            # NOTE rotated files will not require write_lock
+            with open(QUEUE_FILE, "r") as q:
+                # seek to previous end position
+                q.seek(start)
+                # can't use `for line in q` because it disables `q.tell()`
+                # and doesn't allow for releasing write_lock between line reads
+                # or using formats with framing other than newline
+                while True:
+                    # write_lock ensures we don't read a partial line
+                    # or write to queue immediately before removing it
+                    with write_lock:
+                        line = q.readline()
+                        if not line:
+                            # at EOF, remove queue and break
+                            # ignore FileNotFoundError
+                            # remove position first to prevent corrupted state
+                            _remove_if_found(QUEUE_POSITION_FILE)
+                            _remove_if_found(QUEUE_FILE)
+                            done = True
+                            break
+                    # when this throws an exception, stop flushing
+                    # TODO return 202 for transient exceptions
+                    client.publish(**ujson.loads(line)).result()
+                    # record new file position
+                    end = q.tell()
```

Comment:
I didn't know what `.tell()` did, so I looked it up. Among other things, the docs say

> f.tell() returns an integer giving the file object’s current position in the file represented as number of bytes from the beginning of the file when in binary mode and an opaque number when in text mode.

I'm not sure what they mean by "opaque number". I suspect this is how they guard against text encoding issues in text mode, where one character can be more than one byte. I'm not sure if this is something to worry about or not.

The docs say that you can use the results of `q.tell()` to seek from the beginning of a file in text mode, but that doing so from the middle of the file is undefined behavior. I think that means that you're in the clear here?